src/error.rs

```diff
@@ -1,85 +1,151 @@
-pub struct HttpError {
-    inner: InnerHttpError,
-    method: Method,
-    path: String,
+pub enum ApiError {
+    GatewayDeleted,
 }
 
-impl HttpError {
-    pub fn request_method(&self) -> &Method {
-        &self.method
-    }
+// Response from server
+#[derive(Debug, Deserialize, Serialize)]
+struct ApiServerResponse {
+    code: String,
+}
 
-    pub fn request_path(&self) -> &str {
-        &self.path
+impl<'de> Deserialize<'de> for ApiError {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use heck::KebabCase;
+        let resp = ApiServerResponse::deserialize(deserializer)?;
+        match resp.code.to_kebab_case().as_str() {
+            "gateway-deleted" => Ok(ApiError::GatewayDeleted),
+            _ => Err(serde::de::Error::custom(format!("Couldn't deserialize a: {:?}", resp))),
+        }
     }
+}
 
-    /// If the error comes from `reqwest`, return the inner `reqwest::Error`.
-    pub fn to_reqwest_error(&self) -> Option<&reqwest::Error> {
-        match &self.inner {
-            InnerHttpError::Url(_) | InnerHttpError::Query(_) => None,
-            InnerHttpError::Reqwest(e) => Some(e),
+impl std::fmt::Display for ApiError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ApiError::GatewayDeleted => write!(f, "ApiError : GatewayDeleted"),
         }
     }
+}
+
+/// Checks if response status isn't a success,
+/// and then tries to extract json error description from it
+///
+/// # Returns
+///
+/// Result with untouched [`Response`] or error if the status isn't a success
+pub(crate) async fn verify_response(
+    response: Result<Response, reqwest::Error>,
+    method: Method,
+    path: &str,
+) -> Result<Response, Error> {
+    let method_cp = method.clone();
+    let response = response.map_err(|e| Error::Reqwest(e, ErrorDetails::new(method, path)))?;
+
+    if !response.status().is_success() {
+        let cp = method_cp.clone();
+        return Err(response.json::<ApiError>().await.map_or_else(
+            |e| Error::ErrorDeserialization(e, ErrorDetails::new(method_cp.clone(), path)),
+            |e| Error::Api(e, ErrorDetails::new(cp, path)),
+        ));
+    }
+
+    Ok(response)
+}
 
-    // Url and Query accessors left out because there's no use case yet
+#[derive(Debug, Clone)]
+pub struct ErrorDetails {
+    pub method: Method,
+    pub path: String,
+}
+
+impl ErrorDetails {
+    fn new(method: Method, path: &str) -> Self {
+        Self { method, path: path.to_owned() }
+    }
 }
 
 #[derive(Debug)]
-enum InnerHttpError {
-    Url(url::ParseError),
-    Query(serde_urlencoded::ser::Error),
-    Reqwest(reqwest::Error),
+pub enum Error {
+    Url(url::ParseError, ErrorDetails),
+    Query(serde_urlencoded::ser::Error, ErrorDetails),
+    Reqwest(reqwest::Error, ErrorDetails),
+    Api(ApiError, ErrorDetails),
+    ErrorDeserialization(reqwest::Error, ErrorDetails),
```

Comment:
What do you think about having one variant containing a struct with the details + inner variant here? Something like

```suggestion
    Request(RequestError),
```

with

```rust
struct RequestError {
    pub method: Method,
    pub path: String,
    pub kind: RequestErrorKind,
}

enum RequestErrorKind {
    Url(url::ParseError,
    Query(serde_urlencoded::ser::Error),
    Reqwest(reqwest::Error),
    Api(ApiError),
}
```

Reply:
I think that is better approach, I will implement it

Reply:
I remember why I don't do it in that way.
In `lib.rs` we have two functions.
``` Rust   
 fn application_id(&self) -> Result<Uuid, Error> {
    self.application_id.ok_or(Error::ApplicationIdMissing)
 }

 fn gateway_id(&self) -> Result<Uuid, Error> {
     self.gateway_id.ok_or(Error::GatewayIdMissing)
 }
```
If we will have a RequestError then I need to pass as parameters for these functions `Method` and `path` but these args don't have any relations with application_id and gateway_id. Also making the method and path  as
 Option doesn't seem very convenient for using.

Reply:
I think you're misunderstanding. I was suggesting that `Error` would continue to be an enum, with variants `Request(RequestError)`, `ApplicationIdMissing` and `GatewayIdMissing`.

Reply:
I just tried to parse the error and it seems not very convenient to get an GatewayDeleted
```Rust
if let Error::Request(RequestError { kind , .. }) = er {
      if let RequestErrorKind::Api(ApiError::GatewayDeleted) = kind {
          
      }
}
```
```Rust
if let Error::Api(ApiError::GatewayDeleted, details) = err {
}
```

And if additional details are needed we easily get it from error or could skip it.

Reply:
You can collapse the two `if let`s in the first example:

```rust
if let Error::Request(RequestError {
    kind: RequestErrorKind::Api(ApiError::GatewayDeleted),
    ..
}) = err
{
}
```

It's longer but doesn't introduce extra nesting.

I don't insist on changing it though.

Reply:
Still, it seems more complicated than it needed. I don't think that method and path will be used a lot in error parsing on lumeod side.

Reply:
Alternatively, you can make `path` and `method` optional.

Reply:
Yeah, I could but I don't see a problem with the current solution. Could you please write the pros and cons for both approaches?

Reply:
I think I was just overreacting to the use of a tuple enum variant with more than one field, which is often a bad idea, especially when types like `String` are used for the fields (which don't make it clear what the field represents).

The code inside `error.rs` might become a little bit clearer with the nested approach and it would be easier to access the path or method, but that's less important than matching on it so I agree that not changing this is better.